rust/actyx/ax-core/src/runtime/operation/validation.rs

```diff
@@ -0,0 +1,653 @@
+fn validate_bool(value: &CborValue, bool_refinement: &Option<bool>) -> Result<(), anyhow::Error> {
+    if let Some(value) = value.as_bool() {
+        if let Some(bool_refinement) = bool_refinement {
+            if value == *bool_refinement {
+                Ok(())
+            } else {
+                Err(anyhow::anyhow!(
+                    "type mismatch, expected value to be {}, received {} instead",
+                    bool_refinement,
+                    value
+                ))
+            }
+        } else {
+            Ok(())
+        }
+    } else {
+        Err(anyhow::anyhow!("type mismatch, expected value to be a boolean"))
+    }
+}
+
+/// Check if a CBOR value is an integer or an integer refinement (e.g. `10`).
+fn validate_number(value: &CborValue, number_refinement: &Option<u64>) -> Result<(), anyhow::Error> {
+    if let Some(Number::Int(value)) = value.as_number() {
+        if let Ok(value) = u64::try_from(*value) {
+            if let Some(number_refinement) = number_refinement {
+                if value == *number_refinement {
+                    Ok(())
+                } else {
+                    Err(anyhow::anyhow!(
+                        "type mismatch, expected value to be {}, received {} instead",
+                        number_refinement,
+                        value
+                    ))
+                }
+            } else {
+                Ok(())
+            }
+        } else {
+            Err(anyhow::anyhow!(
+                "type mismatch, expected value to be a 64-bit integer but received a 128-bit integer instead"
+            ))
+        }
+    } else {
+        Err(anyhow::anyhow!("type mismatch, expected value to be an integer"))
+    }
+}
+
+/// Check if a CBOR value is a timestamp.
+fn validate_timestamp(value: &CborValue) -> Result<(), anyhow::Error> {
+    if value.as_timestamp().is_some() {
+        Ok(())
+    } else {
+        Err(anyhow::anyhow!("type mismatch, expected value to be a timestamp"))
+    }
+}
+
+/// Check if a CBOR value is a string or a string refinement (e.g. "Hello").
+fn validate_string(value: &CborValue, string_refinement: &Option<String>) -> Result<(), anyhow::Error> {
+    if let Some(value) = value.as_str() {
+        if let Some(refinement) = string_refinement {
+            if value == refinement {
+                Ok(())
+            } else {
+                Err(anyhow::anyhow!(
+                    "type mismatch, expected value to be {}, received {} instead",
+                    refinement,
+                    value
+                ))
+            }
+        } else {
+            Ok(())
+        }
+    } else {
+        Err(anyhow::anyhow!("type mismatch, expected value to be a string"))
+    }
+}
+
+/// Check if a CBOR value is an array. Can also be used to check for tuples (following RFC 7049).
+fn validate_array(value: &CborValue) -> Result<(), anyhow::Error> {
```

Comment:
I can change to the full path as you showed. Otherwise I find that type aliasing `Result<T, E>` to be an anti pattern. It worsens readability and comprehension to save characters, hence originally avoiding it

Reply:
I agree that `use anyhow::Result` shouldn’t be done, but using the fully qualified type is both explicit and easy to read — arguably easier than `Result<(), anyhow::Error>`.